### DISCUSSION TITLE
[@types/koa__router] Weird 'no overload matches this call' error caused by latest version

### DISCUSSION THREAD

**Discussion by cduff** [DefinitelyTyped/DefinitelyTyped#54534] [Issues with a @types package]
The latest version ([8.0.7](https://www.npmjs.com/package/@types/koa__router/v/8.0.7)) of [@types/koa__router](https://www.npmjs.com/package/@types/koa__router) has caused weird errors in my project. I've traced the change that causes it to [pull request 54209](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/54209) by @sedenardi.

The following basic code in a vanilla TypeScript project can be used to reproduce:

``` TypeScript
import Router, { RouterContext } from "@koa/router";
import { DefaultContext, DefaultState } from "koa";

function getRoot(ctx: RouterContext<DefaultState, DefaultContext>) {}

// This causes error
const router = new Router() as Router<DefaultState, DefaultContext>;

// This does not
//const router = new Router<DefaultState, DefaultContext>();

router.get("/", getRoot);
```

This results in the following TypeScript error at the `getRoot` location in the final line:

```
No overload matches this call.
  Overload 1 of 2, '(name: string, path: string | RegExp, ...middleware: Middleware<DefaultState, DefaultContext>[]): Router<DefaultState, DefaultContext>', gave the following error.
    Argument of type '(ctx: RouterContext<DefaultState, DefaultContext>) => void' is not assignable to parameter of type 'string | RegExp'.
      Type '(ctx: RouterContext<DefaultState, DefaultContext>) => void' is missing the following properties from type 'RegExp': exec, test, source, global, and 13 more.
  Overload 2 of 2, '(path: string | RegExp | (string | RegExp)[], ...middleware: Middleware<DefaultState, DefaultContext & { state: DefaultState; } & { body: unknown; response: { ...; }; }>[]): Router<...>', gave the following error.
    Argument of type '(ctx: RouterContext<DefaultState, DefaultContext>) => void' is not assignable to parameter of type 'Middleware<DefaultState, DefaultContext & { state: DefaultState; } & { body: unknown; response: { body: unknown; }; }>'.
      Types of parameters 'ctx' and 'context' are incompatible.
        Type 'ParameterizedContext<DefaultState, DefaultContext & { state: DefaultState; } & { body: unknown; response: { body: unknown; }; } & RouterParamContext<DefaultState, DefaultContext & { ...; } & { ...; }>, any>' is not assignable to type 'RouterContext<DefaultState, DefaultContext>'.
          Type 'ParameterizedContext<DefaultState, DefaultContext & { state: DefaultState; } & { body: unknown; response: { body: unknown; }; } & RouterParamContext<DefaultState, DefaultContext & { ...; } & { ...; }>, any>' is not assignable to type 'RouterParamContext<DefaultState, DefaultContext>'.
            The types returned by 'router.routes()' are incompatible between these types.
              Type 'Middleware<DefaultState, DefaultContext & { state: DefaultState; } & { body: unknown; response: { body: unknown; }; }>' is not assignable to type 'Middleware<DefaultState, DefaultContext>'.
                Type 'DefaultContext' is not assignable to type 'DefaultContext & { state: DefaultState; } & { body: unknown; response: { body: unknown; }; }'.
                  Type 'DefaultContext' is not assignable to type '{ state: DefaultState; }'.
```

If I swap the `router` construction line of code with the commented out alternative then the error goes away. This is strange as both should result in the same `Router` object with the same type?!

Does this make sense? Maybe I'm not understanding something here?

Version details:
* Node 14.17.3
* typescript 4.3.5
* koa 2.13.1
* @koa/router 10.0.0
* @types/koa 2.13.4
* @types/koa__router 8.0.7

**Comment by sedenardi** [comment]
I'm not able to reproduce getting an error with your code, Typescript compiles it just fine for me.

- Node 14.7.3
- Typescript 4.3.4
- koa 2.13.1
- @koa/router 10.0.0
- @types/koa 2.13.3
- @types/koa__router 8.0.7

**Reply by cduff** [reply]
Hi @sedenardi FYI I've created repo for reproduction: [https://github.com/cduff/koa-router-types-issue](https://github.com/cduff/koa-router-types-issue).

**Reply by sedenardi** [reply]
So if I use this as the function parameter, it works fine.

```ts
function getRoot(ctx: Router.Middleware<DefaultState, DefaultContext>) {}
```

In fact, if I look at the type of `ctx` when using an anonymous function

```ts
router.get('/', async (ctx) => {

});
```

I see the type as

`Application.ParameterizedContext<DefaultState, DefaultContext & Router.RouterParamContext<DefaultState, DefaultContext>, any>`

on both the latest version and previous versions.

I think what happened is that you're use of `RouterContext` worked in older versions because it duck-typed to be equivalent to `Router.Middleware<DefaultState, DefaultContext>`. But if you look at the type definitions, both the latest and older versions, they always specified the type of the parameters as `[Koa|Router].Middleware`, not `RouterContext` (which is only used in `ParamMiddleware`).

`@koa/router` treats all the functions passed in as parameters to the `router.[METHOD]()` functions the same, they're all "middleware" functions. So I think you should be using `Router.Middleware` rather than `RouterContext`.

I hope this helps. It may be worth looping in the original author and listed definition authors if you have more questions, as I may have hit the limit of my knowledge on these type definitions (although I'm happy to discuss how the actual `@koa/router` library works).

**Reply by cduff** [reply]
Thanks for your response @sedenardi.

`Router.Middleware` defines a function of form `(ctx, next) => any` whereas `ctx` is supposed to represent the context object.

It gets really weird if you introduce the following couple of lines:

``` TypeScript
...
const router2 = new Router<DefaultState, DefaultContext>();
router2.get("/", getRoot);

// This causes error
...
```

This should have no impact but it makes the error go away?! Could this therefore be exposing a TypeScript bug?

**Reply by sedenardi** [reply]
Ahh yes my mistake, I meant if I assign `Router.Middleware<DefaultState, DefaultContext>` to `getRoot` itself, not its parameter

```ts
// I think you can only assign types to function expressions,
// not function declarations, but could be wrong
const getRoot2: Router.Middleware<DefaultState, DefaultContext> = function(ctx) { };
```

then your code will work. Alternative, if you don't want to type the function itself but just the parameters, then you could grab the type of the `ctx` parameter from `Router.Middleware<DefaultState, DefaultContext>`, and assign it to your `getRoot` parameter:

```ts
type MiddlewareParams = Parameters<Router.Middleware<DefaultState, DefaultContext>>;
type RouterContext = MiddlewareParams[0];
function getRoot(ctx: RouterContext) {}
```

The main point I was trying to make is that `RouterContext` shouldn't be used as the `ctx` type for the router middleware functions. It isn't used in any of the type definitions for the route-handling middleware functions, regardless of the version of this module.

**Reply by cduff** [reply]
Thanks @sedenardi I've now defined my own context type to use in all my middleware functions:

``` TypeScript
type Ctx = Parameters<Router.Middleware<DefaultState, DefaultContext>>[0];
```

I had thought the `RouterContext` type had been provided in the [@types/koa__router](https://www.npmjs.com/package/@types/koa__router) package for this purpose and I still don't understand how the types differ. They generally seem to be compatible with each other. For example the following compiles fine:

``` TypeScript
import Router, { RouterContext } from "@koa/router";
import { DefaultContext, DefaultState } from "koa";

const a = {} as Parameters<Router.Middleware<DefaultState, DefaultContext>>[0];
const b: RouterContext<DefaultState, DefaultContext> = a;
```

Did you try adding the 2 lines of `router2` code as per my response above which makes the error go away? It's really weird and unless I'm misunderstanding something I think it could point to a TypeScript bug?

**Reply by sedenardi** [reply]
`Koa.Middleware` is typed differently than `Koa.ParameterizedContext`. I doubt it's a Typescript bug, it may in fact be a bug in these types that `RouterContext` exists at all, but I'm not totally sure since I didn't write it and I've never used it.

**Reply by cduff** [reply]
I suspect the behaviour we're seeing is related to the [strictFunctionTypes](https://github.com/microsoft/TypeScript/pull/18654) TypeScript feature and how it doesn't apply to function types originating in method or constructor declarations.
